Framework/DataObjects/src/EventList.cpp

```diff
@@ -2851,45 +2854,57 @@ std::vector<double> EventList::getWeightErrors() const {
   return weightErrors;
 }
 
-// --------------------------------------------------------------------------
-/** Get the pulsetimes of all events in a list
- *
- * @param events :: source vector of events
- * @param times :: vector to fill
- */
-template <class T>
-void EventList::getPulseTimesHelper(const std::vector<T> &events,
-                                    std::vector<Mantid::Types::Core::DateAndTime> &times) {
-  times.clear();
-  times.reserve(events.size());
-  std::transform(events.cbegin(), events.cend(), std::back_inserter(times),
-                 [](const auto &event) { return event.pulseTime(); });
-}
-
-/** Get the pulse times of each event in this EventList.
- *
- * @return by copy a vector of DateAndTime times
+/**
+ * Compute a time (for instance, pulse-time plus TOF) associated to each event in the list.
+ * @param timesCalc : anonymous function that computes a time from an input event
+ * @return : vector of times for all events in the list
  */
-std::vector<Mantid::Types::Core::DateAndTime> EventList::getPulseTimes() const {
-  std::vector<Mantid::Types::Core::DateAndTime> times;
-  // Set the capacity of the vector to avoid multiple resizes
-  times.reserve(this->getNumberEvents());
-
-  // Convert the list
+template <typename UnaryOperation>
+std::vector<DateAndTime> EventList::eventTimesCalculator(const UnaryOperation &timesCalc) const {
+  std::vector<DateAndTime> times;
   switch (eventType) {
   case TOF:
-    this->getPulseTimesHelper(this->events, times);
+    times.reserve(events.size());
+    std::transform(events.cbegin(), events.cend(), std::back_inserter(times), timesCalc);
     break;
   case WEIGHTED:
-    this->getPulseTimesHelper(this->weightedEvents, times);
+    times.reserve(weightedEvents.size());
+    std::transform(weightedEvents.cbegin(), weightedEvents.cend(), std::back_inserter(times), timesCalc);
     break;
   case WEIGHTED_NOTIME:
-    this->getPulseTimesHelper(this->weightedEventsNoTime, times);
+    times.reserve(weightedEventsNoTime.size());
+    std::transform(weightedEventsNoTime.cbegin(), weightedEventsNoTime.cend(), std::back_inserter(times), timesCalc);
     break;
   }
   return times;
 }
 
+/** Get the pulse times of each event in this EventList.
+ *
+ * @return by copy a vector of DateAndTime times
+ */
+std::vector<Mantid::Types::Core::DateAndTime> EventList::getPulseTimes() const {
+  auto timeCalc = [](const auto &event) { return event.pulseTime(); };
+  return eventTimesCalculator(timeCalc);
+}
+
+/// Get the Pulse-time + TOF for each event in this EventList
+std::vector<DateAndTime> EventList::getPulseTOFTimes() const {
+  auto timeCalc = [](const auto &event) { return event.pulseTime() + static_cast<int64_t>(event.tof() * 1000.0); };
```

Comment:
Please add a comment about what the 1000 is for

Reply:
I introduced a `constexpr`
```cpp
constexpr double microToNano{1000.0};  // time unit conversion
```